File: LeetCode/Ruby/merge_two_sorted_lists.rb
require '../LeetCode/list_node'
class MergeTwoSortedLists
  def merge_two_lists(l1, l2)
    dummy = ListNode.new(0)
    merged = ListNode.new(0)
    dummy.next = merged
    while l1 != nil && l2 != nil
      if (l1.val > l2.val)
        merged.next = ListNode.new(l2.val)
        l2 = l2.next
      else
        merged.next = ListNode.new(l1.val)
        l1 = l1.next
      end
      merged = merged.next
    end
    merged.next = l1 != nil ? l1 : merged.next
    merged.next = l2 != nil ? l2 : merged.next
    return dummy.next.next
  end
end

l1 = ListNode.new(1)
l1.next = ListNode.new(2)
l1.next.next = ListNode.new(3)
l2 = ListNode.new(1)
l2.next = ListNode.new(3)
l2.next.next = ListNode.new(5)

test = MergeTwoSortedLists.new().merge_two_lists(l1, nil)
while test != nil
  puts test.val
  test = test.next
end

File: LintCode/MergeTwoSortedLists.java
/**
 * Merge Two Sorted Lists
 * http://www.lintcode.com/en/problem/merge-two-sorted-lists/
 *
 * Merge two sorted(ascending) linked lists and return it as a new sorted
 * list. The new sorted list should be made by splicing together the nodes of
 * the two lists and sorted in ascending order.
 *
 * Example
 * Given 1->3->8->11->15->null, 2->null, return 1->2->3->8->11->15->null.
 */

/**
 * Definition for ListNode.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int val) {
 *         this.val = val;
 *         this.next = null;
 *     }
 * }
 */
public class Solution {
  /**
   * @param ListNode l1 is the head of the linked list
   * @param ListNode l2 is the head of the linked list
   * @return: ListNode head of linked list
   */
  public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
    ListNode dummy = new ListNode(0);
    ListNode curt = dummy;
    while (l1 != null && l2 != null) {
      if (l1.val < l2.val) {
        curt.next = new ListNode(l1.val);
        l1 = l1.next;
      } else {
        curt.next = new ListNode(l2.val);
        l2 = l2.next;
      }
      curt = curt.next;
    }
    if (l1 != null) {
      curt.next = l1;
    }
    if (l2 != null) {
      curt.next = l2;
    }
    return dummy.next;
  }

}
